tests/unit/test_plugin.py

```diff
@@ -51,25 +55,200 @@ def test_setup_adapter_fails(mock_check_metadata_version):
+        }).encode()
+    ]
+    mock_get_msi_token.return_value = "Bearer 123456789"
+
+    urlopen.__enter__.return_value = urlopen
+    mock_urlopen.return_value = urlopen
+    caplog.set_level(logging.INFO)
+
+    dimensions = {'tier_1': 10, 'tier_2': 5}
+    timestamp = datetime.datetime.now(datetime.timezone.utc)
+
+    mock_urlopen.return_value = urlopen
+    usage_id = plugin.meter_billing(
+        config,
+        dimensions,
+        timestamp,
+        dry_run=False
+    )
+    assert usage_id["tier_1"] == {"record_id": "1000", "status": "submitted"}
+    assert "New metered billing record added" in caplog.records[0].msg
+
 
+@patch.dict(os.environ, {'EXTENSION_RESOURCE_ID': 'foo', 'PLAN_ID': 'foo'})
+@patch('csp_billing_adapter_microsoft.plugin._get_msi_token')
+@patch('csp_billing_adapter_microsoft.plugin.urllib.request.urlopen')
+def test_meter_billing_urllib_error(mock_urlopen, mock_get_msi_token):
+    """ Test urllib error"""
+    urlopen = MagicMock()
+    urlopen.read.side_effect = [
+        urllib.error.URLError('Cannot get metadata!'),
+        urllib.error.URLError('Cannot get metadata!'),
+        urllib.error.URLError('Cannot get metadata!')
+    ]
+    mock_get_msi_token.return_value = {
+            "Bearer 123456789"
+    }
 
-def test_meter_billing_error():
-    pass
+    urlopen.__enter__.return_value = urlopen
+    mock_urlopen.return_value = urlopen
+
+    dimensions = {'tier_1': 10}
+    timestamp = datetime.datetime.now(datetime.timezone.utc)
+
+    mock_urlopen.return_value = urlopen
+    usage = plugin.meter_billing(config, dimensions, timestamp, dry_run=False)
+    assert usage["tier_1"] == {
+        'status': 'failed',
+        'error': "Failed to meter bill dimensions {'tier_1': 10}: "
+        "<urlopen error Cannot get metadata!>"
+    }
+
+
+@patch.dict(os.environ, {'EXTENSION_RESOURCE_ID': 'foo', 'PLAN_ID': 'foo'})
+@patch('csp_billing_adapter_microsoft.plugin._get_msi_token')
+@patch('csp_billing_adapter_microsoft.plugin.urllib.request.urlopen')
+def test_meter_billing_quantity_is_zero(
+    mock_urlopen,
+    mock_get_msi_token,
+    caplog
+):
+    """ Test when a 0 is passed as the quantity"""
+    urlopen = Mock()
+    urlopen.read.side_effect = [
+        urllib.error.URLError('Cannot get metadata!'),
+        urllib.error.URLError('Cannot get metadata!'),
+        urllib.error.URLError('Cannot get metadata!')
+    ]
+    mock_get_msi_token.return_value = {
+            "Bearer 123456789"
+    }
+
+    mock_urlopen.return_value = urlopen
+    caplog.set_level(logging.INFO)
+
+    dimensions = {'tier_1': 0}
+    timestamp = datetime.datetime.now(datetime.timezone.utc)
+
+    mock_urlopen.return_value = urlopen
+    assert plugin.meter_billing(
+        config,
+        dimensions,
+        timestamp,
+        dry_run=False
+    ) == {}
+
+    assert 'A "0" value was reported for' in caplog.records[0].msg
+    assert "Nothing to meter bill: No tiers have non zero quantity values" in \
```

Comment:
The tiers here would need to change to dimensions too:
```suggestion
    assert "Nothing to meter bill: No dimensions have non zero quantity values" in \
```

Reply:
fixed